R/replace_na.R

```diff
@@ -36,29 +35,50 @@ replace_na <- function(data, replace, ...) {
 #' @export
 replace_na.default <- function(data, replace = NA, ...) {
   check_replacement(replace, "data")
-  data[!is_complete(data)] <- replace
-  data
+  missing <- vec_equal_na(data)
+  vec_assign(data, missing, replace, x_arg = "data", value_arg = "replace")
 }
 
 #' @export
 replace_na.data.frame <- function(data, replace = list(), ...) {
-  stopifnot(is_list(replace))
+  if (!vec_is_list(replace)) {
+    abort("`replace` must be a list.")
+  }
+
+  names <- intersect(names(replace), names(data))
+
+  col_args <- as.character(glue("data${names}"))
+  value_args <- as.character(glue("replace${names}"))
+
+  for (i in seq_along(names)) {
+    name <- names[[i]]
 
-  replace_vars <- intersect(names(replace), names(data))
+    col <- data[[name]]
+    value <- replace[[name]]
 
-  for (var in replace_vars) {
-    check_replacement(replace[[var]], var)
-    data[[var]][!is_complete(data[[var]])] <- replace[[var]]
+    col_arg <- col_args[[i]]
+    value_arg <- value_args[[i]]
+
+    check_replacement(value, col_arg)
```

Comment:
Do we still need this? Or can we rely on `vec_assign()` to apply the recycling rules appropriately?

Reply:
I think we still need this, just because of the way `replace_na()` works.

Otherwise the `replace` value would have to be either size 1 or the same size as `sum(is.na(data))`, and that is almost never the case because you don't know how many missing values there are ahead of time.

```r
x <- c(1, 2, NA, 3, NA)

# This works, of course
replace_na(x, 0)
#> [1] 1 2 0 3 0

# This would actually work too if we removed the restriction
# (because there are two missing values)
# (I don't think we want this to work)
replace_na(x, c(0, -1))
#> [1] 1 2 0 3 -1

# This would not work
# (because replace_na() wants to replace with something of length 2)
y <- c(1, 2, 3, 4, 5)
replace_na(x, y)
# Can't recycle `replace` (size 5) to size 2.
```